src/Core/Lower.hs

```diff
@@ -50,101 +48,78 @@ errRef = CotRef (TgInternal "error")
-
-          fixup (CotyRows rho []) = rho
-          fixup x = x
-
-          pat = CopExtend (CopCapture rest (fixup realt)) [(key, CopCapture var t')]
-          ref = CotRef var t'
-      CotMatch <$> lowerExpr ex <*> pure [(pat, ext, ref)]
-    Tuple xs _ -> do
-      let go :: MonadLower m => Int -> Expr Typed -> m (T.Text, CoType, CoTerm)
-          go k x = (,,) <$> pure (T.pack (show k))
-                        <*> lowerType (getType x)
-                        <*> lowerExpr x
-      CotExtend (CotLit ColRecNil) <$> zipWithM go [1..] xs
-    TypeApp f x _ -> CotTyApp <$> lowerExpr f <*> lowerType x
-    x -> error $ "impossible lowering (desugarer removes): " ++ show (pretty x)
+lowerExpr e = lowerAt e =<< lowerType (getType e)
+
+lowerAt :: MonadLower m => Expr Typed -> CoType -> m CoTerm
+lowerAt e (CotyForall vs b) = flip (foldr (\x -> CotLam Big (x, CotyStar))) vs <$> lowerAt e b
+lowerAt (VarRef (TvName p) _) ty = pure (CotRef p ty)
+lowerAt (Let vs t _) ty = do
+  vs' <- for vs $ \(TvName var, ex, (_, ty)) -> do
+    ty' <- lowerType ty
+    (,,) var <$> pure ty' <*> lowerAt ex ty'
+  CotLet vs' <$> lowerAt t ty
+lowerAt (If c t e _) ty = do
+  (c', t', e') <- (,,) <$> lowerAt c cotyBool <*> lowerAt t ty <*> lowerAt e ty
+  let tc = (CopLit ColTrue, cotyBool, t')
+      te = (CopLit ColFalse, cotyBool, e')
+  pure (CotMatch c' [tc, te])
+lowerAt (App f x _) _ = -- type ignored, ughr
+  CotApp <$> lowerExpr f <*> lowerExpr x
+lowerAt (Fun p bd an) (CotyArr a b) =
+  let operational (PType p _ _) = operational p
+      operational p = p
+   in case operational p of
+        Capture (TvName v) _ -> CotLam Small (v, a) <$> lowerAt bd b
+        _ -> do
+          (p', bd') <- (,) <$> lowerPat p <*> lowerAt bd b
+          arg <- fresh
+          fail <- patternMatchingFail (fst an) a
+          pure (CotLam Small (arg, a) (CotMatch (CotRef arg a) [ (p', a, bd'), fail ]))
+lowerAt (Begin [x] _) t = lowerAt x t
+lowerAt (Begin xs _) t = CotBegin <$> traverse lowerExpr (init xs) <*> lowerAt (last xs) t
+lowerAt (Match ex cs an) ty = do
+  mt <- lowerType (getType ex)
+  cs' <- for cs $ \(pat, ex) ->
+    (,,) <$> lowerPat pat <*> pure mt <*> lowerAt ex ty
+  fail <- patternMatchingFail (fst an) mt
+  CotMatch <$> lowerAt ex mt <*> pure (cs' ++ [fail])
+lowerAt (BinOp left op right a) t = lowerAt (App (App op left a) right a) t
+lowerAt Hole{} _ = error "holes can't be lowered"
+lowerAt (Ascription e t _) _ = lowerAt e =<< lowerType t
+lowerAt e _ = lowerAnyway e
+
+lowerAnyway :: MonadLower m => Expr Typed -> m CoTerm
+lowerAnyway (Record xs _) = case xs of
+  [] -> pure (CotLit ColRecNil)
+  xs -> do
+    xs' <- for xs $ \(label, ex) ->
+      (,,) <$> pure label <*> lowerType (getType ex) <*> lowerExpr ex
+    pure (CotExtend (CotLit ColRecNil) xs')
+lowerAnyway (RecordExt e xs _) = do
+  xs' <- for xs $ \(label, ex) ->
+    (,,) <$> pure label <*> lowerType (getType ex) <*> lowerExpr ex
```

Comment:
Tuple sections? I know you don't like them, so fair doos if not.

Reply:
Eh, sure.